test/blackbox-tests/utils/webserver_oneshot.ml

```diff
@@ -0,0 +1,61 @@
+(* An http server which serves the contents of a given file a single time and then terminates *)
+
+module Args = struct
+  type t =
+    { content_file : string
+    ; port_file : string
+    ; simulate_not_found : bool
+    }
+
+  let parse () =
+    let content_file = ref "" in
+    let port_file = ref "" in
+    let simulate_not_found = ref false in
+    let specs =
+      [ "--content-file", Arg.Set_string content_file, "File to serve"
+      ; ( "--port-file"
+        , Arg.Set_string port_file
+        , "The server will write its port number to this file" )
+      ; "--simulate-not-found", Arg.Set simulate_not_found, "Return a 404 page"
+      ]
+    in
+    Arg.parse
+      specs
+      (fun _anon_arg -> failwith "unexpected anonymous argument")
+      "Run a webserver on a random port which serves the contents of a  single file a \
+       single time, then terminates.";
+    { content_file = !content_file
+    ; port_file = !port_file
+    ; simulate_not_found = !simulate_not_found
+    }
+  ;;
+end
+
+let main content_file port_file ~simulate_not_found =
+  let host = Unix.inet_addr_loopback in
+  let addr = Unix.ADDR_INET (host, 0) in
+  let sock = Unix.socket ~cloexec:true Unix.PF_INET Unix.SOCK_STREAM 0 in
+  Unix.setsockopt sock Unix.SO_REUSEADDR true;
+  Unix.bind sock addr;
+  Unix.listen sock 1;
+  let port =
+    match Unix.getsockname sock with
+    | Unix.ADDR_INET (_, port) -> port
+    | ADDR_UNIX _ -> failwith "unreachable"
+  in
+  Out_channel.with_open_text port_file (fun out_channel ->
+    Out_channel.output_string out_channel (Printf.sprintf "%d" port));
```

Comment:
Probably would make sense to add a final newline here, this way it doesn't mess up terminals when `cat`-ed and capturing it with `$()` should still work.